audit/base/base.go

```diff
@@ -0,0 +1,556 @@
+	UnknownErrorCategory ErrorCategory = 0
+	// ErrValidation is an error that occurred during validation of an audit record.
+	// This means the audit record is invalid and was not sent.
+	ErrValidation ErrorCategory = 1
+	// ErrConnection is an error that occurred while connecting to the audit server.
+	// This means the audit client is in an unrecoverable state. Getting this error means that every
+	// other call to the client will fail with the same message.
+	ErrConnection ErrorCategory = 2
+	// ErrQueueFull is an error that occurred because the queue is full.
+	ErrQueueFull ErrorCategory = 3
+	// ErrTimeout is an error that occurred because the send timed out.
+	ErrTimeout ErrorCategory = 4
+)
+
+// Error represents an error that occurred in this package. All errors returned by this package
+// are of type Error.
+type Error struct {
+	// Err is the underlying error.
+	Err error
+	// Category is the category of the error.
+	Category ErrorCategory
+}
+
+// Error returns the error message.
+func (e Error) Error() string {
+	return fmt.Sprintf("%s: %s", e.Category, e.Err.Error())
+}
+
+// Unwrap returns the underlying error.
+func (e Error) Unwrap() error {
+	return e.Err
+}
+
+// IsUnrecoverable returns true if the error is unrecoverable. This only works on
+// errors of type Error (from this package). If not Error, this returns false. Unrecoverable errors
+// mean that the audit client should not attempt to send any more audit records as the
+// client is in an unrecoverable state (e.g. the connection is dead).
+func IsUnrecoverable(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if e, ok := err.(Error); ok {
+		return e.Category == ErrConnection
+	}
+
+	return false
+}
+
+// IsQueueFull returns true if the error is because the queue is full.
+func IsQueueFull(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if e, ok := err.(Error); ok {
+		return e.Category == ErrQueueFull
+	}
+
+	return false
+}
+
+// IsValidationErr returns true if the error is because the audit record is invalid.
+func IsValidationErr(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if e, ok := err.(Error); ok {
+		return e.Category == ErrValidation
+	}
+
+	return false
+}
+
+// Client represents a client connection to an audit server.
+// Note that there should be only one client created per application. We do not use a singleton to prevent this,
+// but rather we rely on the user to only create one client.
+// Note: Context cancellation is only supported if explicitly noted in the method documentation.
+type Client struct {
+	// conn is the connection to the audit server.
+	conn atomic.Pointer[conn.Audit]
+	// settings are the settings for when the audit client sends audit records.
+	settings Settings
+	// sendCh is the channel for sending audit records to our async sender. It has a buffer
+	// capacity of MaxQueueSize.
+	sendCh      chan streamData[SendMsg]
+	stopSend    chan chan struct{}
+	successSend bool
```

Comment:
```suggestion
	// successSend tracks if we've successfully sent at least one message to the audit server. This is so that we can
	// start the heartbeat only once a successful message has been sent
	successSend bool
```